api/v1alpha1/snapshotenvironmentbinding_webhook.go

```diff
@@ -30,61 +32,113 @@ import (
+	}
+
+	snapshotenvironmentbindinglog := snapshotenvironmentbindinglog.WithValues("controllerKind", "SnapshotEnvironmentBinding").WithValues("name", binding.Name).WithValues("namespace", binding.Namespace)
+	snapshotenvironmentbindinglog.Info("default", "name", binding.Name)
+
+	return nil
 }
 
 // change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
-//+kubebuilder:webhook:path=/validate-appstudio-redhat-com-v1alpha1-snapshotenvironmentbinding,mutating=false,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=snapshotenvironmentbindings,verbs=create;update,versions=v1alpha1,name=vsnapshotenvironmentbinding.kb.io,admissionReviewVersions=v1
+// +kubebuilder:webhook:path=/validate-appstudio-redhat-com-v1alpha1-snapshotenvironmentbinding,mutating=false,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=snapshotenvironmentbindings,verbs=create;update,versions=v1alpha1,name=vsnapshotenvironmentbinding.kb.io,admissionReviewVersions=v1
+type snapshotEnvironmentBindingWebhookHandler struct {
+	client.Client
+}
 
-var _ webhook.Validator = &SnapshotEnvironmentBinding{}
+var _ webhook.CustomValidator = &snapshotEnvironmentBindingWebhookHandler{}
+
+func (h *snapshotEnvironmentBindingWebhookHandler) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	binding, ok := obj.(*SnapshotEnvironmentBinding)
+	if !ok {
+		return nil, fmt.Errorf("runtime object is not of type SnapshotEnvironmentBinding")
+	}
 
-// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
-func (r *SnapshotEnvironmentBinding) ValidateCreate() (admission.Warnings, error) {
-	snapshotenvironmentbindinglog := snapshotenvironmentbindinglog.WithValues("controllerKind", "SnapshotEnvironmentBinding").WithValues("name", r.Name).WithValues("namespace", r.Namespace)
+	snapshotenvironmentbindinglog := snapshotenvironmentbindinglog.WithValues("controllerKind", "SnapshotEnvironmentBinding").WithValues("name", binding.Name).WithValues("namespace", binding.Namespace)
 	snapshotenvironmentbindinglog.Info("validating create")
 
+	// Retrieve the list of existing SnapshotEnvironmentBindings from the namespace
+	existingSEBs := SnapshotEnvironmentBindingList{}
+
+	if err := h.Client.List(context.Background(), &existingSEBs, &client.ListOptions{Namespace: binding.Namespace}); err != nil {
+		return nil, fmt.Errorf("failed to list existing SnapshotEnvironmentBindings: %v", err)
+	}
+
+	// Check if any existing SEB has the same Application/Environment combination
+	for _, existingSEB := range existingSEBs.Items {
+		if existingSEB.Spec.Application == binding.Spec.Application && existingSEB.Spec.Environment == binding.Spec.Environment {
+			return nil, fmt.Errorf("duplicate combination of Application (%s) and Environment (%s)", binding.Spec.Application, binding.Spec.Environment)
+		}
+	}
```

Comment:
This bit of code is the same between Create and Update AFAICT, and thus could be moved into its own function to reduce code dupe.